core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,56 +125,76 @@ public class JsPeripheral internal constructor(
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            disconnectJob.launchAsync()
             throw t
         }
-
-        logger.info { message = "Connected" }
-        _state.value = State.Connected
     }
 
-    private fun closeConnection() {
-        observationListeners.clear()
-        disconnectGatt()
-        unregisterDisconnectedListener()
-        _state.value = State.Disconnected()
+    private val disconnectJob: SharedRepeatableTask = scope.sharedRepeatableTask {
+        connectJob.join()
+        closeConnection()
     }
 
     public override suspend fun connect() {
-        val job = connectJob ?: connectAsync().also { connectJob = it }
-        job.await()
+        connectJob.launchAsync().await()
     }
 
     public override suspend fun disconnect() {
-        job.cancelAndJoinChildren()
-        connectJob = null
-        disconnectGatt()
+        connectJob.cancel()
+        disconnectJob.launchAsync().await()
+    }
+
+    private suspend fun openConnection() {
+        // Guard if already connected do nothing
+        if (_state.value is State.Connected) return
+
+        logger.info { message = "Connecting" }
+        // Wait until fully disconnected before proceeding
+        _state.first { it is State.Disconnected }
+
+        _state.value = State.Connecting.Bluetooth
+        registerDisconnectedListener()
+        gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
+
+        _state.value = State.Connecting.Services
+        discoverServices()
+        onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
+
+        _state.value = State.Connecting.Observes
+        logger.verbose { message = "Configuring characteristic observations" }
+        observers.onConnected()
+
+        logger.info { message = "Connected" }
+        _state.value = State.Connected
     }
 
-    private fun disconnectGatt() {
-        _state.value = State.Disconnecting
-        bluetoothDevice.gatt?.disconnect()
-        // Avoid trampling existing `Disconnected` state (and its properties) by only updating if not already `Disconnected`.
-        _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
+    private suspend fun closeConnection() {
+        // Guard if already disconnecting/disconnected do nothing
+        if (_state.value is State.Disconnecting || _state.value is State.Disconnected) return
+
+        try {
+            _state.value = State.Disconnecting
+            unregisterDisconnectedListener()
+            stopAllObservations()
+            bluetoothDevice.gatt?.disconnect()
```

Comment:
Should this be moved into `finally` block? to make sure that we cleanup underlying BLE connection even if `stopAllObservations` fails?

Reply:
I guess if we move it to the `finally` block and ensure that the state is updated _before_ invoking it in case it fails that may be the best approach.

Reply:
Oh, I missed that.
I was wanting to move it to the `finally` block to make sure that a disconnect occurs even if stopping observations failed.

Maybe keeping the state update after the disconnect is appropriate? If `bluetoothDevice.gatt?.disconnect()` fails, should we be updating to `Disconnected` state? 🤷

Reply:
Yes we should. These are the places where we are guaranteeing that we are moving to the expected state regardless if the underlying system is borked. 

If `bluetoothDevice.gatt?.disconnect()` if broken we can be certain the connection is broken and need to be in the `Disconnected` state.